### PR TITLE
Add short syntax for generative functor types: () -> ...

### PR DESCRIPTION
**Pull request by yallop**
#42 added short syntax for non-dependent functors.  A non-dependent functor

```ocaml
module A (X : t) : t = N
```

has the following type

```ocaml
t -> t
```

Separately, #5905 added generative functors to OCaml.  A generative functor 

```ocaml
module G () : t = N
```
has the following more verbose type

```ocaml
functor () -> t
```

For consistency and conciseness this little PR adds short syntax for generative functors, too, so that `G` can be given the type

```ocaml
() -> t
```

### CODE REVIEW

**Comment by gasche** [comment]
This sounds reasonable to me. Note that the new syntax is not fully "consistent" with the non-dependent functor case: in the non-dependent case we abbreviate `functor (_ : S1) -> S2` as `S1 -> S2`, so we use the signature of the input on the left of the arrow. For generative functors, `()` is ad-hoc syntax that is neither a module parameter name nor the signature of a module parameter.

We could think of `()` as the only inhabitant of a module-level `Unit` type (which does not contain structures), and `functor () -> M` as pattern-matching on the inhabitant `()` at abstraction time. In this view the "natural" non-dependent functor syntax would be `Unit -> S` rather than `() -> S`, and the use of `() -> S` corresponds to a pun on the type of unit (as if we wrote `()` for both the unit value and the unit type, at the level of terms).

Long story short: this is not the same thing as the current non-dependent functor syntax, nor a canonical extension of it. It is a separate syntax proposal for a shorter signature for (non-dependent, obviously) generative functors. This proposal looks reasonable to me.

**Comment by yallop** [comment]
One way to think of `()` in functor parameters and types is to consider it a shorthand for `() : ()`, but with an obligation to omit the redundant type constituent when writing terms and to omit the redundant term constituent when writing types.  Under that interpretation the existing generative functor syntax doesn't seem so ad-hoc and the proposed syntax is consistent with the existing short syntax.

**Review by nojb — APPROVED** [review]
LGTM

As this is a syntax change, a second approval would be safer (even if perhaps not strictly necessary as per current policy). @gasche: you have already looked at the PR, do you mind approving it officially (if you agree with it)? Thanks!

Incidentally, the PR updates the vendored version of `menhirLib`; I don't remember if we had a policy regarding this or not...

**Comment by garrigue** [comment]
I had the same reaction as @gasche, that this is not really symmetric.
But since there is no predefined `Unit` type, this is not shocking to write it `()`, as this is a very special case.
So this looks fine to me too.

**Comment by nojb** [comment]
Also forgot: a `Changes` entry is in order, as well as adapting the syntax in the manual.

**Comment by yallop** [comment]
@nojb:
> Also forgot: a Changes entry is in order, as well as adapting the syntax in the manual.

Done: [ac94d28](https://github.com/ocaml/ocaml/pull/11694/commits/ac94d289ca6bd8afe57ada0c086bfa787dddf45e), [3a38583](https://github.com/ocaml/ocaml/pull/11694/commits/3a38583f0ab7b834da4409425730160a4d0c4c33) (respectively)

**Comment by nojb** [comment]
> @nojb:
> 
> > Also forgot: a Changes entry is in order, as well as adapting the syntax in the manual.
> 
> Done: [ac94d28](https://github.com/ocaml/ocaml/pull/11694/commits/ac94d289ca6bd8afe57ada0c086bfa787dddf45e), [3a38583](https://github.com/ocaml/ocaml/pull/11694/commits/3a38583f0ab7b834da4409425730160a4d0c4c33) (respectively)

Thanks. The PR needs to be rebased on `trunk`, and then we can move to merge it.

**Comment by yallop** [comment]
> Thanks. The PR needs to be rebased on `trunk`, and then we can move to merge it.

Squashed and rebased (and re-promoted with menhir 20210419 to avoid forcing everyone to upgrade): ~~[62d7b25](https://github.com/ocaml/ocaml/pull/11694/commits/62d7b252a0ceaa88853c7324a44f7e7801ffe14f)~~ [043edc4](https://github.com/ocaml/ocaml/pull/11694/commits/043edc49c93ced84a0fbeeb5e4a721e36add65b1) (now with a couple of fixes for changed syntax error messages in the test)

**Comment by nojb** [comment]
> > Thanks. The PR needs to be rebased on `trunk`, and then we can move to merge it.
> 
> Squashed and rebased (and re-promoted with menhir 20210419 to avoid forcing everyone to upgrade): [62d7b25](https://github.com/ocaml/ocaml/pull/11694/commits/62d7b252a0ceaa88853c7324a44f7e7801ffe14f)

Thanks! This looks good to merge to me.

**Comment by gasche** [comment]
(For the record, my approach is that upgrading to a newer version of Menhir is always fine.)
